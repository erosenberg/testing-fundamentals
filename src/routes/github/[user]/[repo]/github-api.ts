import type { paths } from "@octokit/openapi-types";
type OrgRepoResponse =
  paths["/repos/{owner}/{repo}"]["get"]["responses"]["200"]["content"]["application/json"];

export type Fetch = typeof fetch;
export class GithubApi {
  constructor(
    private token: string | undefined,
    private fetch: Fetch,
    private delay: (ms: number) => Promise<void>
  ) {}

  async getRepositories(username: string) {
    let page = 1;
    const repositories: OrgRepoResponse[] = [];

    while (true) {
      const response = await this.fetch(
        `https://api.github.com/users/${username}/repos?per_page=30&page=${page}`,
        {
          headers: {
            "User-Agent": "Qwik Workshop",
            "X-GitHub-Api-Version": "2022-11-28",
          },
        }
      );
      const json = await response.json();
      repositories.push(...json);
      console.log(json);
      console.log(repositories);

      if (json.length < 30) {
        break;
      }
      page++;
    }
    return repositories;
  }

  async getRepository(user: string, repo: string) {
    const headers: HeadersInit = {
      "User-Agent": "Qwik Workshop",
      "X-GitHub-Api-Version": "2022-11-28",
    };
    if (this.token) {
      headers["Authorization"] = "Bearer " + this.token;
    }

    return Promise.race([
      this.delay(4000).then(() => {
        return {
          response: "timeout",
        };
      }),
      this.fetch(`https://api.github.com/repos/${user}/${repo}`, {
        headers,
      }).then((response) => {
        return response.json();
      }),
    ]);
  }
}

export function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
