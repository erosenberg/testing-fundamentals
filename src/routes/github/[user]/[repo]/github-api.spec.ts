import { describe, it, expect } from "vitest";
import { GithubApi } from "./github-api";

// Filename
describe("github-api", () => {
  // Function name
  describe("getRepository", () => {
    // Every 'it' is a Product Requirement
    it("should return repository information", async () => {
      const api = new GithubApi(undefined);
      const response = await api.getRepository("mhevery", "qwik");
      expect(response).toMatchSnapshot();
    });
    it.todo("should return timeout after x seconds with time out response");
  });
});
