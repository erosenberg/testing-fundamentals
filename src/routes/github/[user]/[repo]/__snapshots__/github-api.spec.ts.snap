// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`github-api > getRepository > should return repository information 1`] = `
{
  "allow_forking": true,
  "archive_url": "https://api.github.com/repos/mhevery/qwik/{archive_format}{/ref}",
  "archived": false,
  "assignees_url": "https://api.github.com/repos/mhevery/qwik/assignees{/user}",
  "blobs_url": "https://api.github.com/repos/mhevery/qwik/git/blobs{/sha}",
  "branches_url": "https://api.github.com/repos/mhevery/qwik/branches{/branch}",
  "clone_url": "https://github.com/mhevery/qwik.git",
  "collaborators_url": "https://api.github.com/repos/mhevery/qwik/collaborators{/collaborator}",
  "comments_url": "https://api.github.com/repos/mhevery/qwik/comments{/number}",
  "commits_url": "https://api.github.com/repos/mhevery/qwik/commits{/sha}",
  "compare_url": "https://api.github.com/repos/mhevery/qwik/compare/{base}...{head}",
  "contents_url": "https://api.github.com/repos/mhevery/qwik/contents/{+path}",
  "contributors_url": "https://api.github.com/repos/mhevery/qwik/contributors",
  "created_at": "2021-06-01T21:33:00Z",
  "default_branch": "main",
  "deployments_url": "https://api.github.com/repos/mhevery/qwik/deployments",
  "description": "An Open-Source framework designed for best possible [time to interactive](https://web.dev/interactive/), by focusing on [resumability](./docs/RESUMABLE.md) of server-side-rendering of HTML, and [fine-grained lazy-loading](./docs/LAZY_LOADING.md) of code.",
  "disabled": false,
  "downloads_url": "https://api.github.com/repos/mhevery/qwik/downloads",
  "events_url": "https://api.github.com/repos/mhevery/qwik/events",
  "fork": true,
  "forks": 1,
  "forks_count": 1,
  "forks_url": "https://api.github.com/repos/mhevery/qwik/forks",
  "full_name": "mhevery/qwik",
  "git_commits_url": "https://api.github.com/repos/mhevery/qwik/git/commits{/sha}",
  "git_refs_url": "https://api.github.com/repos/mhevery/qwik/git/refs{/sha}",
  "git_tags_url": "https://api.github.com/repos/mhevery/qwik/git/tags{/sha}",
  "git_url": "git://github.com/mhevery/qwik.git",
  "has_discussions": false,
  "has_downloads": true,
  "has_issues": false,
  "has_pages": false,
  "has_projects": true,
  "has_wiki": true,
  "homepage": null,
  "hooks_url": "https://api.github.com/repos/mhevery/qwik/hooks",
  "html_url": "https://github.com/mhevery/qwik",
  "id": 372966483,
  "is_template": false,
  "issue_comment_url": "https://api.github.com/repos/mhevery/qwik/issues/comments{/number}",
  "issue_events_url": "https://api.github.com/repos/mhevery/qwik/issues/events{/number}",
  "issues_url": "https://api.github.com/repos/mhevery/qwik/issues{/number}",
  "keys_url": "https://api.github.com/repos/mhevery/qwik/keys{/key_id}",
  "labels_url": "https://api.github.com/repos/mhevery/qwik/labels{/name}",
  "language": "TypeScript",
  "languages_url": "https://api.github.com/repos/mhevery/qwik/languages",
  "license": {
    "key": "mit",
    "name": "MIT License",
    "node_id": "MDc6TGljZW5zZTEz",
    "spdx_id": "MIT",
    "url": "https://api.github.com/licenses/mit",
  },
  "merges_url": "https://api.github.com/repos/mhevery/qwik/merges",
  "milestones_url": "https://api.github.com/repos/mhevery/qwik/milestones{/number}",
  "mirror_url": null,
  "name": "qwik",
  "network_count": 1227,
  "node_id": "MDEwOlJlcG9zaXRvcnkzNzI5NjY0ODM=",
  "notifications_url": "https://api.github.com/repos/mhevery/qwik/notifications{?since,all,participating}",
  "open_issues": 0,
  "open_issues_count": 0,
  "owner": {
    "avatar_url": "https://avatars.githubusercontent.com/u/111951?v=4",
    "events_url": "https://api.github.com/users/mhevery/events{/privacy}",
    "followers_url": "https://api.github.com/users/mhevery/followers",
    "following_url": "https://api.github.com/users/mhevery/following{/other_user}",
    "gists_url": "https://api.github.com/users/mhevery/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/mhevery",
    "id": 111951,
    "login": "mhevery",
    "node_id": "MDQ6VXNlcjExMTk1MQ==",
    "organizations_url": "https://api.github.com/users/mhevery/orgs",
    "received_events_url": "https://api.github.com/users/mhevery/received_events",
    "repos_url": "https://api.github.com/users/mhevery/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/mhevery/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mhevery/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/mhevery",
  },
  "parent": {
    "allow_forking": true,
    "archive_url": "https://api.github.com/repos/QwikDev/qwik/{archive_format}{/ref}",
    "archived": false,
    "assignees_url": "https://api.github.com/repos/QwikDev/qwik/assignees{/user}",
    "blobs_url": "https://api.github.com/repos/QwikDev/qwik/git/blobs{/sha}",
    "branches_url": "https://api.github.com/repos/QwikDev/qwik/branches{/branch}",
    "clone_url": "https://github.com/QwikDev/qwik.git",
    "collaborators_url": "https://api.github.com/repos/QwikDev/qwik/collaborators{/collaborator}",
    "comments_url": "https://api.github.com/repos/QwikDev/qwik/comments{/number}",
    "commits_url": "https://api.github.com/repos/QwikDev/qwik/commits{/sha}",
    "compare_url": "https://api.github.com/repos/QwikDev/qwik/compare/{base}...{head}",
    "contents_url": "https://api.github.com/repos/QwikDev/qwik/contents/{+path}",
    "contributors_url": "https://api.github.com/repos/QwikDev/qwik/contributors",
    "created_at": "2021-05-19T15:33:32Z",
    "default_branch": "main",
    "deployments_url": "https://api.github.com/repos/QwikDev/qwik/deployments",
    "description": "Instant-loading web apps, without effort",
    "disabled": false,
    "downloads_url": "https://api.github.com/repos/QwikDev/qwik/downloads",
    "events_url": "https://api.github.com/repos/QwikDev/qwik/events",
    "fork": false,
    "forks": 1227,
    "forks_count": 1227,
    "forks_url": "https://api.github.com/repos/QwikDev/qwik/forks",
    "full_name": "QwikDev/qwik",
    "git_commits_url": "https://api.github.com/repos/QwikDev/qwik/git/commits{/sha}",
    "git_refs_url": "https://api.github.com/repos/QwikDev/qwik/git/refs{/sha}",
    "git_tags_url": "https://api.github.com/repos/QwikDev/qwik/git/tags{/sha}",
    "git_url": "git://github.com/QwikDev/qwik.git",
    "has_discussions": false,
    "has_downloads": true,
    "has_issues": true,
    "has_pages": false,
    "has_projects": true,
    "has_wiki": false,
    "homepage": "https://qwik.dev",
    "hooks_url": "https://api.github.com/repos/QwikDev/qwik/hooks",
    "html_url": "https://github.com/QwikDev/qwik",
    "id": 368917877,
    "is_template": false,
    "issue_comment_url": "https://api.github.com/repos/QwikDev/qwik/issues/comments{/number}",
    "issue_events_url": "https://api.github.com/repos/QwikDev/qwik/issues/events{/number}",
    "issues_url": "https://api.github.com/repos/QwikDev/qwik/issues{/number}",
    "keys_url": "https://api.github.com/repos/QwikDev/qwik/keys{/key_id}",
    "labels_url": "https://api.github.com/repos/QwikDev/qwik/labels{/name}",
    "language": "TypeScript",
    "languages_url": "https://api.github.com/repos/QwikDev/qwik/languages",
    "license": {
      "key": "mit",
      "name": "MIT License",
      "node_id": "MDc6TGljZW5zZTEz",
      "spdx_id": "MIT",
      "url": "https://api.github.com/licenses/mit",
    },
    "merges_url": "https://api.github.com/repos/QwikDev/qwik/merges",
    "milestones_url": "https://api.github.com/repos/QwikDev/qwik/milestones{/number}",
    "mirror_url": null,
    "name": "qwik",
    "node_id": "MDEwOlJlcG9zaXRvcnkzNjg5MTc4Nzc=",
    "notifications_url": "https://api.github.com/repos/QwikDev/qwik/notifications{?since,all,participating}",
    "open_issues": 363,
    "open_issues_count": 363,
    "owner": {
      "avatar_url": "https://avatars.githubusercontent.com/u/138123704?v=4",
      "events_url": "https://api.github.com/users/QwikDev/events{/privacy}",
      "followers_url": "https://api.github.com/users/QwikDev/followers",
      "following_url": "https://api.github.com/users/QwikDev/following{/other_user}",
      "gists_url": "https://api.github.com/users/QwikDev/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/QwikDev",
      "id": 138123704,
      "login": "QwikDev",
      "node_id": "O_kgDOCDuZuA",
      "organizations_url": "https://api.github.com/users/QwikDev/orgs",
      "received_events_url": "https://api.github.com/users/QwikDev/received_events",
      "repos_url": "https://api.github.com/users/QwikDev/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/QwikDev/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/QwikDev/subscriptions",
      "type": "Organization",
      "url": "https://api.github.com/users/QwikDev",
    },
    "private": false,
    "pulls_url": "https://api.github.com/repos/QwikDev/qwik/pulls{/number}",
    "pushed_at": "2024-05-07T17:00:51Z",
    "releases_url": "https://api.github.com/repos/QwikDev/qwik/releases{/id}",
    "size": 42063,
    "ssh_url": "git@github.com:QwikDev/qwik.git",
    "stargazers_count": 20266,
    "stargazers_url": "https://api.github.com/repos/QwikDev/qwik/stargazers",
    "statuses_url": "https://api.github.com/repos/QwikDev/qwik/statuses/{sha}",
    "subscribers_url": "https://api.github.com/repos/QwikDev/qwik/subscribers",
    "subscription_url": "https://api.github.com/repos/QwikDev/qwik/subscription",
    "svn_url": "https://github.com/QwikDev/qwik",
    "tags_url": "https://api.github.com/repos/QwikDev/qwik/tags",
    "teams_url": "https://api.github.com/repos/QwikDev/qwik/teams",
    "topics": [
      "framework",
      "javascript",
      "web",
    ],
    "trees_url": "https://api.github.com/repos/QwikDev/qwik/git/trees{/sha}",
    "updated_at": "2024-05-07T16:48:37Z",
    "url": "https://api.github.com/repos/QwikDev/qwik",
    "visibility": "public",
    "watchers": 20266,
    "watchers_count": 20266,
    "web_commit_signoff_required": false,
  },
  "private": false,
  "pulls_url": "https://api.github.com/repos/mhevery/qwik/pulls{/number}",
  "pushed_at": "2024-04-25T00:52:57Z",
  "releases_url": "https://api.github.com/repos/mhevery/qwik/releases{/id}",
  "size": 40211,
  "source": {
    "allow_forking": true,
    "archive_url": "https://api.github.com/repos/QwikDev/qwik/{archive_format}{/ref}",
    "archived": false,
    "assignees_url": "https://api.github.com/repos/QwikDev/qwik/assignees{/user}",
    "blobs_url": "https://api.github.com/repos/QwikDev/qwik/git/blobs{/sha}",
    "branches_url": "https://api.github.com/repos/QwikDev/qwik/branches{/branch}",
    "clone_url": "https://github.com/QwikDev/qwik.git",
    "collaborators_url": "https://api.github.com/repos/QwikDev/qwik/collaborators{/collaborator}",
    "comments_url": "https://api.github.com/repos/QwikDev/qwik/comments{/number}",
    "commits_url": "https://api.github.com/repos/QwikDev/qwik/commits{/sha}",
    "compare_url": "https://api.github.com/repos/QwikDev/qwik/compare/{base}...{head}",
    "contents_url": "https://api.github.com/repos/QwikDev/qwik/contents/{+path}",
    "contributors_url": "https://api.github.com/repos/QwikDev/qwik/contributors",
    "created_at": "2021-05-19T15:33:32Z",
    "default_branch": "main",
    "deployments_url": "https://api.github.com/repos/QwikDev/qwik/deployments",
    "description": "Instant-loading web apps, without effort",
    "disabled": false,
    "downloads_url": "https://api.github.com/repos/QwikDev/qwik/downloads",
    "events_url": "https://api.github.com/repos/QwikDev/qwik/events",
    "fork": false,
    "forks": 1227,
    "forks_count": 1227,
    "forks_url": "https://api.github.com/repos/QwikDev/qwik/forks",
    "full_name": "QwikDev/qwik",
    "git_commits_url": "https://api.github.com/repos/QwikDev/qwik/git/commits{/sha}",
    "git_refs_url": "https://api.github.com/repos/QwikDev/qwik/git/refs{/sha}",
    "git_tags_url": "https://api.github.com/repos/QwikDev/qwik/git/tags{/sha}",
    "git_url": "git://github.com/QwikDev/qwik.git",
    "has_discussions": false,
    "has_downloads": true,
    "has_issues": true,
    "has_pages": false,
    "has_projects": true,
    "has_wiki": false,
    "homepage": "https://qwik.dev",
    "hooks_url": "https://api.github.com/repos/QwikDev/qwik/hooks",
    "html_url": "https://github.com/QwikDev/qwik",
    "id": 368917877,
    "is_template": false,
    "issue_comment_url": "https://api.github.com/repos/QwikDev/qwik/issues/comments{/number}",
    "issue_events_url": "https://api.github.com/repos/QwikDev/qwik/issues/events{/number}",
    "issues_url": "https://api.github.com/repos/QwikDev/qwik/issues{/number}",
    "keys_url": "https://api.github.com/repos/QwikDev/qwik/keys{/key_id}",
    "labels_url": "https://api.github.com/repos/QwikDev/qwik/labels{/name}",
    "language": "TypeScript",
    "languages_url": "https://api.github.com/repos/QwikDev/qwik/languages",
    "license": {
      "key": "mit",
      "name": "MIT License",
      "node_id": "MDc6TGljZW5zZTEz",
      "spdx_id": "MIT",
      "url": "https://api.github.com/licenses/mit",
    },
    "merges_url": "https://api.github.com/repos/QwikDev/qwik/merges",
    "milestones_url": "https://api.github.com/repos/QwikDev/qwik/milestones{/number}",
    "mirror_url": null,
    "name": "qwik",
    "node_id": "MDEwOlJlcG9zaXRvcnkzNjg5MTc4Nzc=",
    "notifications_url": "https://api.github.com/repos/QwikDev/qwik/notifications{?since,all,participating}",
    "open_issues": 363,
    "open_issues_count": 363,
    "owner": {
      "avatar_url": "https://avatars.githubusercontent.com/u/138123704?v=4",
      "events_url": "https://api.github.com/users/QwikDev/events{/privacy}",
      "followers_url": "https://api.github.com/users/QwikDev/followers",
      "following_url": "https://api.github.com/users/QwikDev/following{/other_user}",
      "gists_url": "https://api.github.com/users/QwikDev/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/QwikDev",
      "id": 138123704,
      "login": "QwikDev",
      "node_id": "O_kgDOCDuZuA",
      "organizations_url": "https://api.github.com/users/QwikDev/orgs",
      "received_events_url": "https://api.github.com/users/QwikDev/received_events",
      "repos_url": "https://api.github.com/users/QwikDev/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/QwikDev/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/QwikDev/subscriptions",
      "type": "Organization",
      "url": "https://api.github.com/users/QwikDev",
    },
    "private": false,
    "pulls_url": "https://api.github.com/repos/QwikDev/qwik/pulls{/number}",
    "pushed_at": "2024-05-07T17:00:51Z",
    "releases_url": "https://api.github.com/repos/QwikDev/qwik/releases{/id}",
    "size": 42063,
    "ssh_url": "git@github.com:QwikDev/qwik.git",
    "stargazers_count": 20266,
    "stargazers_url": "https://api.github.com/repos/QwikDev/qwik/stargazers",
    "statuses_url": "https://api.github.com/repos/QwikDev/qwik/statuses/{sha}",
    "subscribers_url": "https://api.github.com/repos/QwikDev/qwik/subscribers",
    "subscription_url": "https://api.github.com/repos/QwikDev/qwik/subscription",
    "svn_url": "https://github.com/QwikDev/qwik",
    "tags_url": "https://api.github.com/repos/QwikDev/qwik/tags",
    "teams_url": "https://api.github.com/repos/QwikDev/qwik/teams",
    "topics": [
      "framework",
      "javascript",
      "web",
    ],
    "trees_url": "https://api.github.com/repos/QwikDev/qwik/git/trees{/sha}",
    "updated_at": "2024-05-07T16:48:37Z",
    "url": "https://api.github.com/repos/QwikDev/qwik",
    "visibility": "public",
    "watchers": 20266,
    "watchers_count": 20266,
    "web_commit_signoff_required": false,
  },
  "ssh_url": "git@github.com:mhevery/qwik.git",
  "stargazers_count": 10,
  "stargazers_url": "https://api.github.com/repos/mhevery/qwik/stargazers",
  "statuses_url": "https://api.github.com/repos/mhevery/qwik/statuses/{sha}",
  "subscribers_count": 2,
  "subscribers_url": "https://api.github.com/repos/mhevery/qwik/subscribers",
  "subscription_url": "https://api.github.com/repos/mhevery/qwik/subscription",
  "svn_url": "https://github.com/mhevery/qwik",
  "tags_url": "https://api.github.com/repos/mhevery/qwik/tags",
  "teams_url": "https://api.github.com/repos/mhevery/qwik/teams",
  "temp_clone_token": null,
  "topics": [],
  "trees_url": "https://api.github.com/repos/mhevery/qwik/git/trees{/sha}",
  "updated_at": "2024-04-18T23:37:43Z",
  "url": "https://api.github.com/repos/mhevery/qwik",
  "visibility": "public",
  "watchers": 10,
  "watchers_count": 10,
  "web_commit_signoff_required": false,
}
`;
